### PR TITLE
[Core][CoSim] Adding size check in Auxiliar Model Part Utilities

### PR DESCRIPTION
**Pull request by ashishdarekar**
**Description**
This PR is to improve the AuxiliarModelPartUtilities

- KratosCore
- AuxiliarModelPartUtitlities
- ParallelUtilities

**Changelog**
- Included the Size Check for SetVector() and SetScalar() Utilities
- Ported For loops to Parallel Utilities

### CODE REVIEW

File: kratos/utilities/auxiliar_model_part_utilities.h [hunk 1]
```diff
@@ -343,40 +351,58 @@ class KRATOS_API(KRATOS_CORE) AuxiliarModelPartUtilities
...
+            TSize = mrModelPart.GetCommunicator().GetDataCommunicator().MaxAll(TSize);
+
             data.resize(mrModelPart.NumberOfElements()*TSize);
+
             GetVectorDataFromContainer(mrModelPart.Elements(), TSize, rVariable, data);
             break;
         }
         case (DataLocation::Condition):{
-            const std::size_t TSize = mrModelPart.NumberOfConditions() > 0 ? mrModelPart.ConditionsBegin()->GetValue(rVariable).size() : 0;
+            unsigned int TSize = mrModelPart.NumberOfConditions() > 0 ? mrModelPart.ConditionsBegin()->GetValue(rVariable).size() : 0;
```
Comment (philbucher) [inline kratos/utilities/auxiliar_model_part_utilities.h:393]:
```suggestion
            std::size_t TSize = mrModelPart.NumberOfConditions() > 0 ? mrModelPart.ConditionsBegin()->GetValue(rVariable).size() : 0;
```

File: kratos/utilities/auxiliar_model_part_utilities.h [hunk 2]
```diff
@@ -412,25 +439,32 @@ class KRATOS_API(KRATOS_CORE) AuxiliarModelPartUtilities
...
             ImportDataSizeCheck(mrModelPart.NumberOfNodes(), rData.size());
-            IndexType counter = 0;
-            for(auto& r_node : mrModelPart.Nodes()){
-                auto& r_val = r_node.FastGetSolutionStepValue(rVariable);
-                r_val = rData[counter++];
-            }
+
+            ModelPart::NodesContainerType::iterator inodebegin = mrModelPart.NodesBegin();
+            IndexPartition<IndexType>(mrModelPart.NumberOfNodes()).for_each([&](IndexType Index){
+                ModelPart::NodesContainerType::iterator inode = inodebegin + Index;
```
Comment (philbucher) [inline kratos/utilities/auxiliar_model_part_utilities.h]:
```suggestion
            auto inodebegin = mrModelPart.NodesBegin();
            IndexPartition<IndexType>(mrModelPart.NumberOfNodes()).for_each([&](IndexType Index){
                auto inode = inodebegin + Index;
```

here `auto` makes the code more readable IMO

**Review by philbucher — COMMENTED** [review]
please use `std::size_t` over `unsigned int`
std::size_t is 64bit on x64 systems and x86 bit on 32bt systems, whereas unsigned int is always 32 bit, which might not be enough for large models!

once again I am quite sure abt the sizes, but for being absolutely sure one should probably double check it

**Comment by philbucher** [comment]
from https://en.cppreference.com/w/cpp/types/size_t:

~~~
std::size_t is commonly used for array indexing and loop counting. Programs that use other types, such as unsigned int, for array indexing may fail on, e.g. 64-bit systems when the index exceeds UINT_MAX or if it relies on 32-bit modular arithmetic. 
~~~

this is what I meant

File: kratos/utilities/auxiliar_model_part_utilities.h (hunk 2 above)
Comment (ashishdarekar) [reply kratos/utilities/auxiliar_model_part_utilities.h]: True!

File: kratos/utilities/auxiliar_model_part_utilities.h (hunk 1 above)
Comment (ashishdarekar) [reply kratos/utilities/auxiliar_model_part_utilities.h:393]: Got it . Thanks :)

**Comment by ashishdarekar** [comment]
> D:\a\Kratos\Kratos\kratos\utilities/auxiliar_model_part_utilities.h(395,1): message : while trying to match the argument list '(size_t)' [D:\a\Kratos\Kratos\build\Custom\kratos\KratosCore.vcxproj]
D:\a\Kratos\Kratos\kratos\utilities/auxiliar_model_part_utilities.h(499,1): error C2668: 'Kratos::DataCommunicator::MaxAll': ambiguous call to overloaded function [D:\a\Kratos\Kratos\build\Custom\kratos\KratosCore.vcxproj]
D:\a\Kratos\Kratos\kratos\includes/data_communicator.h(356,5): message : could be 'double Kratos::DataCommunicator::MaxAll(const double) const' [D:\a\Kratos\Kratos\build\Custom\kratos\KratosCore.vcxproj]
D:\a\Kratos\Kratos\kratos\includes/data_communicator.h(355,5): message : or       'unsigned long Kratos::DataCommunicator::MaxAll(const unsigned long) const' [D:\a\Kratos\Kratos\build\Custom\kratos\KratosCore.vcxproj]
D:\a\Kratos\Kratos\kratos\includes/data_communicator.h(354,5): message : or       'unsigned int Kratos::DataCommunicator::MaxAll(const unsigned int) const' [D:\a\Kratos\Kratos\build\Custom\kratos\KratosCore.vcxproj]
D:\a\Kratos\Kratos\kratos\includes/data_communicator.h(353,5): message : or       'int Kratos::DataCommunicator::MaxAll(const int) const' [D:\a\Kratos\Kratos\build\Custom\kratos\KratosCore.vcxproj]
D:\a\Kratos\Kratos\kratos\utilities/auxiliar_model_part_utilities.h(499,1): message : while trying to match the argument list '(size_t)' [D:\a\Kratos\Kratos\build\Custom\kratos\KratosCore.vcxproj]
D:/a/Kratos/Kratos/kratos/tests/cpp_tests/utilities/test_auxiliar_model_part_utilities.cpp(481): message : see reference to function template instantiation 'void Kratos::AuxiliarModelPartUtilities::SetVectorData<Kratos::array_1d<double,3>>(const Kratos::Variable<Kratos::array_1d<double,3>> &,const Kratos::DataLocation,const std::vector<double,std::allocator<T>> &)' being compiled [D:\a\Kratos\Kratos\build\Custom\kratos\KratosCore.vcxproj]
          with
          [
              T=double
          ]

Windows Failing: 
@philbucher Yes, I remembered now. To avoid above error only, I have changed the Datatype of Tsize variable to **unsigned int** 
Maybe I am wrong or there might be a different solution to this.

**Comment by philbucher** [comment]
ah ok I understand, sorry I forgot about that.
Then lets go with `unsigned int` for now!

**Review by philbucher — APPROVED** [review]

**Pull request merged**